### PR TITLE
[jit] support serialization of classes

### PR DESCRIPTION
**Pull request by suo**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#17856 [jit] support serialization of classes**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14402599/)

Add support for saving/loading TorchScript modules that depend on user-defned classes.

We track class dependencies the same we track tensor constants, then write them
all out such that we can just compile them in order before compiling the module
hierarchy.

Differential Revision: [D14402599](https://our.internmc.facebook.com/intern/diff/D14402599/)

### CODE REVIEW

File: torch/csrc/jit/import.cpp [hunk 1]
```diff
@@ -138,6 +140,21 @@ void ScriptModuleDeserializer::loadTensorTable(torch::ModelDef* model_def) {
   }
 }
 
+void ScriptModuleDeserializer::loadLibs(torch::ModelDef* model_def) {
+  for (const auto& lib_def : model_def->libs()) {
+    auto module = std::make_shared<script::Module>();
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/jit/import.cpp]: What happens to module? It just goes away? Is import_methods actually loading a class into the global namespace here?

**Review by zdevito — COMMENTED** [review]
This looks pretty good. My only substantive comments are around how classes are re-registered (I am not sure I understand why we are returning seemingly unused modules), and about trying to get all the class definitions into one easy-to-read file.

File: torch/csrc/jit/import.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (suo) [reply torch/csrc/jit/import.cpp]: Yeah, the module gets owned by the class type which is generated when importing methods. It's a consequence of the fact that class types are registered globally as a side effect of compilation. Not sure what to do about it besides make it more clear what's happening?

File: torch/csrc/jit/passes/python_print.cpp [hunk 2]
```diff
@@ -883,10 +906,18 @@ struct PythonPrintPass {
             [graph, name, this] { printFunctionDefinition(*graph, name); });
         stmt << "self." << name;
       } break;
-      case prim::CreateObject:
-      case prim::SetAttr:
-      case prim::GetAttr:
-        throw std::runtime_error("NYI");
+      case prim::CreateObject: {
+        const auto classType = node->output()->type()->expect<ClassType>();
+        addToClassTable(classType);
```
[the remark this answers is not in the document]
Comment (suo) [reply torch/csrc/jit/passes/python_print.cpp]: Well if you construct a class object in the method and only use it internally (maybe for bookkeeping or something) then we still need the definition. We can also examine the type of `useOf` and add it there I suppose?

[the remark this answers is not in the document]
Comment (suo) [reply torch/csrc/jit/passes/python_print.cpp]: Oh, you were saying each node, not the graph in general. Makes sense

File: torch/csrc/jit/script/sugared_value.h [hunk 3]
```diff
@@ -298,6 +303,35 @@ struct TORCH_API IsInstanceValue : SugaredValue {
   }
 };
 
+// This represents the "__new__" method on classes, which can't be a MethodValue
```
[the remark this answers is not in the document]
Comment (suo) [reply torch/csrc/jit/script/sugared_value.h]: wouldn't you need a a "createObject" sugared value and a dynamic_cast anyway? (similar to how AnnotateValue and ForkValue are implemented).

At least this way we also get the __new__ syntax as well.

[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/jit/script/sugared_value.h]: Yeah, I see that it doesn't solve anything now. Let's leave it as is. I am going to figure out if we can add a new sugared method that systematizes the places where we are dynamic casting the sugared values.

File: torch/csrc/jit/passes/python_print.cpp [hunk 4]
```diff
@@ -643,6 +655,19 @@ struct PythonPrintPass {
   }
 
   void printNode(Node* node, bool print_const) {
+    // Check for class dependencies. If this node inputs or outputs a class
```
Comment (zdevito) [inline torch/csrc/jit/passes/python_print.cpp]: What about `List[Class]` or `Tuple`?

**Review by zdevito — APPROVED** [review]
Looks good. One minor issue.

File: torch/csrc/jit/passes/python_print.cpp (hunk 4 above)
Comment (suo) [reply torch/csrc/jit/passes/python_print.cpp]: aahh good point.

File: torch/csrc/jit/import_source.cpp [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    env_ = {
+        {"torch", std::make_shared<BuiltinModule>("aten", version)},
+        {"ops", std::make_shared<OpsValue>(version)},
+        {"CONSTANTS", std::make_shared<ConstantTableValue>(constant_table)},
+        {"fork", std::make_shared<ForkValue>()},
+        {"annotate", std::make_shared<AnnotateValue>()},
+        {"inf",
+         std::make_shared<ConstantValue>(
+             std::numeric_limits<double>::infinity())},
+        {"nan",
```
Comment (eellison) [inline torch/csrc/jit/import_source.cpp]: Why was "nan" added here?

Comment (suo) [reply torch/csrc/jit/import_source.cpp]: no clue. It may have been a bad merge :/

**Pull request closed without merging**